pkg/shortnames/shortnames.go

```diff
@@ -0,0 +1,396 @@
+// returned as is.  In case, it's a short name, it's resolved according to the
+// ShortNameMode in the SystemContext (if specified) or in the registries.conf.
+//
+// Note that tags and digests are stripped from the specified name before
+// looking up an alias.  The returned slice may be empty if no matching alias
+// and no unqualified-search registries could be found.
+
+// Note that callers *must* call `(Resolve).Resolved()` after a returned item
+// has been pulled successfully; this callback will record a new short-name
+// alias (depending on the specified short-name mode).
+func Resolve(ctx *types.SystemContext, name string) ([]Resolved, error) {
+	// Create a copy of the system context to make it useable beyond this
+	// function call.
+	var sys *types.SystemContext
+	if ctx != nil {
+		sys = &(*ctx)
+	}
+
+	// Detect which mode we're running in.
+	mode, err := sysregistriesv2.GetShortNameMode(sys)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sanity check the short-name mode.
+	switch mode {
+	case types.ShortNameModeDisabled, types.ShortNameModePermissive, types.ShortNameModeEnforcing:
+		// We're good.
+	default:
+		return nil, errors.Errorf("unsupported short-name mode (%v)", mode)
+	}
+
+	isShort, shortRef, err := isShortName(name)
+	if err != nil {
+		return nil, err
+	}
+	if !isShort { // no short name
+		named, isNamed := shortRef.(reference.Named)
+		if !isNamed {
+			return nil, errors.Wrapf(err, "name %q is neither a short name nor a valid named reference", name)
+		}
+		named = reference.TagNameOnly(named) // Make sure to add ":latest" if needed
+		return []Resolved{newResolved(name, named, false, sys)}, nil
+	}
+
+	// Strip off the tag to normalize the short name for looking it up in
+	// the config files.
+	isTagged, isDigested, normalizedShortName, tag, digest := normalizeUserInput(shortRef)
+
+	// If there's already an alias, use it.
+	isAliased, namedAlias, err := sysregistriesv2.ResolveShortNameAlias(sys, normalizedShortName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Always use an alias if present.
+	if isAliased {
+		// If the input was tagged, add it to the alias.
+		if isTagged {
+			namedAlias, err = reference.WithTag(namedAlias, tag)
+			if err != nil {
+				return nil, err
+			}
+		} else if isDigested {
+			namedAlias, err = reference.WithDigest(namedAlias, digest)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			// Make sure to add ":latest" if needed
+			namedAlias = reference.TagNameOnly(namedAlias)
+		}
+		return []Resolved{newResolved(name, namedAlias, false, sys)}, nil
+	}
+
+	// Query the registry for unqualified-search registries.
+	unqualifiedSearchRegistries, err := sysregistriesv2.UnqualifiedSearchRegistries(sys)
+	if err != nil {
+		return nil, err
+	}
+
+	var candidates []Resolved
+	for _, reg := range unqualifiedSearchRegistries {
+		named, err := reference.ParseNormalizedNamed(fmt.Sprintf("%s/%s", reg, name))
```

Comment:
(Absolutely non-blocking: In both cases of this loop: This is of course correct, but a tiny bit non-obvious that this code path can ignore `isTagged`/`isDigested`/`tag`/`digest`. Consolidating/factoring out the re-tagging/re-digesting and using it consistently on all paths might be worth considering. OTOH this is short and simple.)